src/tpm2_dictionarylockout.c

```diff
@@ -0,0 +1,229 @@
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// 3. Neither the name of Intel Corporation nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <stdbool.h>
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <getopt.h>
+
+#include <sapi/tpm20.h>
+
+#include "log.h"
+#include "main.h"
+#include "options.h"
+#include "string-bytes.h"
+
+#define PASSWORD_MAX (sizeof(((TPM2B_DIGEST *)NULL)->t.buffer))
+
+typedef struct dictionarylockout_ctx dictionarylockout_ctx;
+struct dictionarylockout_ctx {
+    UINT32 max_tries;
+    UINT32 recovery_time;
+    UINT32 lockout_recovery;
+    char lockout_passwd[PASSWORD_MAX];
+    bool clear_lockout;
+    bool setup_parameters;
+    bool use_passwd;
+    TSS2_SYS_CONTEXT *sapi_context;
+};
+
+bool dictionary_lockout_reset_and_parameter_setup(dictionarylockout_ctx *ctx) {
+    
+    TPMS_AUTH_COMMAND sessionData;
+    TPMS_AUTH_RESPONSE sessionDataOut;
+    TSS2_SYS_CMD_AUTHS sessionsData;
+    TSS2_SYS_RSP_AUTHS sessionsDataOut;
+    TPMS_AUTH_COMMAND *sessionDataArray[1];
+    TPMS_AUTH_RESPONSE *sessionDataOutArray[1];
+
+    sessionDataArray[0] = &sessionData;
+    sessionDataOutArray[0] = &sessionDataOut;
+    sessionsDataOut.rspAuths = &sessionDataOutArray[0];
+    sessionsData.cmdAuths = &sessionDataArray[0];
+    sessionsDataOut.rspAuthsCount = 1;
+    sessionData.sessionHandle = TPM_RS_PW;
+    sessionData.nonce.t.size = 0;
+    sessionData.hmac.t.size = 0;
+    *( (UINT8 *)((void *)&sessionData.sessionAttributes ) ) = 0;
```

Comment:
So this prologue is a copy paste nightmare that I finally have some semblance of understanding on. The cast to a UINT8 pointer and then de-referencing it and clearing it to 0 is a poor optimization that could go wrong. the sessionAtrributes field is a bitfiled of 8 used buts on U32 union. They attempted to just set those bits without writing to the whole U32. Just set the whole U32 to 0 in case the flags ever change...